{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ACGGCTAG', 'ACGTATGC', 'ATG', 'TAGC']\n",
      "['TAGC', 'ACGTATGC', 'ATG', 'ACGGCTAG']\n",
      "['ACGTATGC', 'ACGGCTAG', 'TAGC', 'ATG']\n",
      "['ACGGCTAG', 'TAGC', 'ACGTATGC', 'ATG']\n",
      "['ACGG', 'ATCGA', 'ATCGAA', 'CGTAAA']\n",
      "[('gene3', 1036, 1966), ('gene4', 543, 522), ('gene1', 121, 98), ('gene2', 56, 32)]\n",
      "[(2, 8766, 'gene4'), (4, 7633, 'gene3'), (4, 9200, 'gene1'), (6, 63788, 'gene2')]\n",
      "['CGATG', 'AATTCTATCGATATCGTGA']\n",
      "['CGATGAATTCTATCGAT', 'ATCGTGA']\n"
     ]
    }
   ],
   "source": [
    "# examples\n",
    "\n",
    "dna_list = ['TAGC', 'ACGTATGC', 'ATG', 'ACGGCTAG']\n",
    "long_dna = []\n",
    "for dna in dna_list:\n",
    "    if len(dna) > 5:\n",
    "        long_dna.append(dna)\n",
    "\n",
    "dna_list = ['TAGC', 'ACGTATGC', 'ATG', 'ACGGCTAG']\n",
    "def is_long(dna):\n",
    "    return len(dna) > 5\n",
    "long_dna = filter(is_long, dna_list)\n",
    "\n",
    "dna_list = ['TAGC', 'ACGTATGC', 'ATG', 'ACGGCTAG']\n",
    "sorted_dna = sorted(dna_list)\n",
    "print(sorted_dna)\n",
    "print(dna_list)\n",
    "\n",
    "#sorted_dna = sorted(dna_list, key=len)\n",
    "sorted_dna = sorted(dna_list, key=len, reverse=True)\n",
    "\n",
    "print(sorted_dna)\n",
    "\n",
    "\n",
    "def get_at(dna):\n",
    "    return (dna.count('A') + dna.count('T')) / len(dna)\n",
    "\n",
    "dna_list = ['TAGC', 'ACGTATGC', 'ATG', 'ACGGCTAG']\n",
    "sorted_dna = sorted(dna_list, key=get_at)\n",
    "print(sorted_dna)\n",
    "\n",
    "import re\n",
    "def poly_a_length(dna):\n",
    "    poly_a_match = re.search(r'A+$', dna)\n",
    "    if poly_a_match:\n",
    "        return len(poly_a_match.group())\n",
    "    else:\n",
    "        return 0\n",
    "\n",
    "dna_list = ['ATCGA', 'ACGG', 'CGTAAA', 'ATCGAA']\n",
    "print(sorted(dna_list, key=poly_a_length))\n",
    "\n",
    "measurements = [\n",
    "('gene1', 121, 98),\n",
    "('gene2', 56, 32),\n",
    "('gene3', 1036, 1966),\n",
    "('gene4', 543, 522)\n",
    "]\n",
    "\n",
    "def get_ratio(measurement):\n",
    "    return measurement[2] / measurement[1]\n",
    "\n",
    "print(sorted(measurements, key=get_ratio, reverse=True))\n",
    "\n",
    "\n",
    "loci = [\n",
    "(4, 9200, 'gene1'),\n",
    "(6, 63788, 'gene2'),\n",
    "(4, 7633, 'gene3'),\n",
    "(2, 8766, 'gene4')\n",
    "]\n",
    "\n",
    "def get_chromosome(locus):\n",
    "    return locus[0]\n",
    "def get_base_position(locus):\n",
    "    return locus[1]\n",
    "\n",
    "sorted_by_base = sorted(loci, key=get_base_position)\n",
    "final_sort = sorted(sorted_by_base, key=get_chromosome)\n",
    "print(final_sort)\n",
    "\n",
    "\n",
    "def get_kmers_at(dna, k):\n",
    "    result = []\n",
    "    for i in range(len(dna) - k +1):\n",
    "        kmer = dna[i:i+k]\n",
    "        at = (kmer.count('A') + kmer.count('T')) / k\n",
    "        result.append(at)\n",
    "    return result\n",
    "\n",
    "def get_kmers_cg(dna, k):\n",
    "    result = []\n",
    "    for i in range(len(dna) - k +1):\n",
    "        kmer = dna[i:i+k]\n",
    "        cg = kmer.count('CG')\n",
    "        result.append(cg)\n",
    "    return result\n",
    "\n",
    "def get_kmers_f(dna, k, analyze_kmer):\n",
    "    result = []\n",
    "    for i in range(len(dna) - k +1):\n",
    "        kmer = dna[i:i+k]\n",
    "        result.append(analyze_kmer(kmer))\n",
    "    return result\n",
    "\n",
    "dna = 'ATCGATCATCGGCATCGATCGGTATCAGTACGTAC'\n",
    "at_scores = get_kmers_f(dna, 8, get_at)\n",
    "\n",
    "cg_counts = get_kmers_f(dna, 8, lambda dna : dna.count('CG'))\n",
    "\n",
    "\n",
    "def make_digester(pattern, offset):\n",
    "    \n",
    "    def digester(dna):\n",
    "        current_position = 0\n",
    "        result = []\n",
    "        for m in re.finditer(pattern, dna):\n",
    "            result.append(dna[current_position:m.start() + offset])\n",
    "            current_position = m.start() + offset\n",
    "        result.append(dna[current_position:])\n",
    "        return result\n",
    "    \n",
    "    return digester\n",
    "\n",
    "dna='CGATGAATTCTATCGATATCGTGA'\n",
    "\n",
    "ecori_digester = make_digester('GAATTC', 1)\n",
    "print(ecori_digester(dna))\n",
    "ecorv_digester = make_digester('GATATC', 3)\n",
    "print(ecorv_digester(dna))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exercises"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gi|336287915|gb|AEI30246.1|\n",
      "gi|336287919|gb|AEI30248.1|\n",
      "gi|336287881|gb|AEI30229.1|\n",
      "gi|336287897|gb|AEI30237.1|\n",
      "gi|336287895|gb|AEI30236.1|\n",
      "gi|336287917|gb|AEI30247.1|\n",
      "gi|336287921|gb|AEI30249.1|\n",
      "gi|336287923|gb|AEI30250.1|\n",
      "gi|336287885|gb|AEI30231.1|\n",
      "gi|336287889|gb|AEI30233.1|\n",
      "0.02262443438914027\n",
      "0.009009009009009009\n",
      "0.02262443438914027\n",
      "0.02262443438914027\n",
      "0.02262443438914027\n",
      "0.02262443438914027\n",
      "0.007797270955165692\n",
      "0.04308390022675737\n"
     ]
    }
   ],
   "source": [
    "# Blast processor\n",
    "def mismatch_filter(hit_string):\n",
    "    mismatch_count = int(hit_string.split(\"\\t\")[4])\n",
    "    return mismatch_count < 20\n",
    "\n",
    "def get_subject(hit_string):\n",
    "    return hit_string.split(\"\\t\")[1]\n",
    "\n",
    "def comment_filter(line):\n",
    "    return not line.startswith('#')\n",
    "\n",
    "def get_percent_id(hit_string):\n",
    "    return float(hit_string.split(\"\\t\")[2])\n",
    "\n",
    "def get_query(hit_string):\n",
    "    return hit_string.split(\"\\t\")[1]\n",
    "\n",
    "def cox1_filter(hit_string):\n",
    "    subject = hit_string.split(\"\\t\")[1]\n",
    "    if \"COX1\" in subject:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def start_ratio(hit_string):\n",
    "    query_start = int(hit_string.split(\"\\t\")[6])\n",
    "    hit_length = int(hit_string.split(\"\\t\")[3])\n",
    "    return query_start / hit_length\n",
    "\n",
    "\n",
    "hit_lines = filter(comment_filter, open('examples_and_exercises/Chapter_5/exercises/blast_result.txt'))\n",
    "f = filter(mismatch_filter, hit_lines)\n",
    "s = sorted(hit_lines, key=get_percent_id)\n",
    "low_id_hits = s[0:10]\n",
    "for subject in map(get_query, low_id_hits):\n",
    "    print(subject)\n",
    "\n",
    "hit_lines = filter(comment_filter, open('examples_and_exercises/Chapter_5/exercises/blast_result.txt'))\n",
    "ratio = filter(cox1_filter, hit_lines)\n",
    "for record in map(start_ratio, ratio):\n",
    "    print(record)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FASTA processor\n",
    "import re\n",
    "\n",
    "def to_upper(dna):\n",
    "    return dna.upper()\n",
    "\n",
    "def do_nothing(x):\n",
    "    return x\n",
    "\n",
    "def remove(dna):\n",
    "    return re.sub(r'[^ATGCatgc]', '', dna)\n",
    "\n",
    "def fix_spaces(header):\n",
    "    return header.replace(' ', '_')\n",
    "\n",
    "def truncate(header):\n",
    "    return header[0:10]\n",
    "\n",
    "def fasta_copy(source, destination, process_header, process_sequence):\n",
    "    output = open(destination, \"w\")\n",
    "    header = \"\"\n",
    "    for line in open(source):\n",
    "        if line.startswith('>'):\n",
    "            header = process_header(line.rstrip(\"\\n\")[1:])\n",
    "        else:\n",
    "            sequence = process_sequence(line.rstrip(\"\\n\"))\n",
    "            output.write('>' + header +\"\\n\")\n",
    "            output.write(sequence + \"\\n\")\n",
    "\n",
    "fasta_copy('examples_and_exercises/Chapter_5/exercises/sequences.fasta', 'corrected.fasta', do_nothing, to_upper)\n",
    "fasta_copy('examples_and_exercises/Chapter_5/exercises/sequences.fasta', 'corrected.fasta', do_nothing, remove)\n",
    "fasta_copy('examples_and_exercises/Chapter_5/exercises/sequences.fasta', 'corrected.fasta', fix_spaces, do_nothing)\n",
    "fasta_copy('examples_and_exercises/Chapter_5/exercises/sequences.fasta', 'corrected.fasta', truncate, do_nothing)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
